test/integration/fake_upstream.cc

```diff
@@ -354,14 +355,24 @@ AssertionResult FakeConnectionBase::close(std::chrono::milliseconds timeout) {
 }
 
 AssertionResult FakeConnectionBase::readDisable(bool disable, std::chrono::milliseconds timeout) {
-  return shared_connection_.executeOnDispatcher(
-      [disable](Network::Connection& connection) { connection.readDisable(disable); }, timeout);
+  if (shared_connection_.connection().dispatcher().isThreadSafe()) {
+    shared_connection_.connection().readDisable(disable);
+    return AssertionSuccess();
+  } else {
+    return shared_connection_.executeOnDispatcher(
+        [disable](Network::Connection& connection) { connection.readDisable(disable); }, timeout);
```

Comment:
executeOnDispatcher blocks on until the posted callback finishes.  Calling post and waiting from the dispatcher thread results in a deadlock.

Reply:
makes sense; adding a code comment may help the reader.

Reply:
Idea: Instead of dealing this in all call sites, could executeOnDispatcher() figure this out and just run the lambda versus posting?

Reply:
I undid this branching and reverted this function.

FakeUpstream now does readDisable and enableHalfClose directly on the network connection instead of going through this wrapper in the cases where the operation happens from the dispatcher thread.  Some tests still call this method and that requires going down the executeOnDispatcher version.